File: epi_judge_python/test_framework/generic_test_handler.py
# @library
import inspect

from test_framework.test_output import TestOutput
from test_framework.test_timer import TestTimer
from test_framework.test_utils_deserialization import get_string_parser_for_type

from test_framework import test_utils


class GenericTestHandler:
    """
    The central class in generic test runner framework.

    It is responsible for constructing string deserializers
    basing on the test data header and
    executing tests on the provided function (which includes
    the deserialization of the provided arguments and the expected value,
    invocation of the target function with these arguments and
    comparison of the computed result with the expected value).

    parse_signature() and run_test() raise RuntimeError
    in case of any error or assertion failure. This exception
    terminates testing and, consequently, the test program.

    If tested function throws TestFailureException or StackOverflowError,
    the current test is raise as failed and the execution goes on.
    In case of any other exception raised by the tested function,
    the test program is terminated.
    """

    def __init__(self, func, comp):
        self._func = func
        self._has_timer_hook = test_utils.has_timer_hook(func)
        self._param_parsers = []
        self._param_names = [
            p.name
            for p in inspect.signature(self._func).parameters.values()
            if p.default is inspect.Parameter.empty
        ][1 if self._has_timer_hook else 0:]
        self._comp = comp

    def parse_signature(self, signature):
        """
        This method initializes type parsers
        for all tested function arguments and expected value,
        basing on test data signature (see get_string_parser_for_type()).

        :param signature: the header from a test data file.
        :type signature: List[str]
        """
        if len(signature) != len(self._param_names) + 1:
            raise RuntimeError("Signature parameter count mismatch")

        signature = [test_utils.filter_bracket_comments(s) for s in signature]

        for param in signature[:-1]:
            self._param_parsers.append(get_string_parser_for_type(param))

        if any(p is None for p in self._param_parsers):
            raise RuntimeError("Argument can't be of type void")

        self._ret_value_parser = get_string_parser_for_type(signature[-1])

    def run_test(self, test_args):
        """
        This method is invoked for each row in a test data file (except the header).
        It deserializes the list of arguments and calls the user function with them.

        :param test_args: serialized arguments
        :return: list, that contains [result of comparison of expected and result, expected, result].
                 Two last entries are omitted in case of the void return type
        """
        args = [
            parser(x)
            for parser, x in zip(self._param_parsers,
                                 test_args[:-1 if not self.expected_is_void()
                                           else len(test_args)])
        ]
        timer = TestTimer()
        if self._has_timer_hook:
            args = [timer] + args

        if not self.expected_is_void():
            expected = self._ret_value_parser(test_args[-1])

            timer.start()
            result = self._func(*args)
            timer.stop()

            return TestOutput(
                self._comp(expected, result), timer, expected, result)
        else:
            timer.start()
            self._func(*args)
            timer.stop()

            return TestOutput(True, timer)

    def expected_is_void(self):
        return self._ret_value_parser is None

    def param_names(self):
        return self._param_names

File: epi_judge_java_solutions/epi/test_framework/GenericTestHandler.java
// @library
package epi.test_framework;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.lang.reflect.Parameter;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.function.BiPredicate;
import java.util.function.Function;
import java.util.stream.Collectors;

/**
 * The central class in generic test runner framework.
 * It is responsible for constructing string deserializers
 * based on the argument types of the tested method (obtained with reflexions)
 * (see {@link #GenericTestHandler(Method, BiPredicate, List)},
 * asserting that the method signature matches the one from the test file header
 * (see {@link #parseSignature(List)} and
 * executing tests on the provided method (which includes
 * the deserialization of the provided arguments and the expected value,
 * invocation of the target method with these arguments and
 * comparison of the computed result with the expected value)
 * (see {@link #runTest(long, List)}).
 * <p>
 * {@link #parseSignature(List)} and {@link #runTest(long, List)} throw {@link
 * RuntimeException} in case of any error or assertion failure. This exception
 * terminates testing and, consequently, the test program. If tested method
 * throws {@link TestFailure} or {@link StackOverflowError}, the
 * current test is marked as failed and the execution goes on. In case of any
 * other exception thrown by the tested method, the test program is terminated.
 * <p>
 */
public class GenericTestHandler {
  private Method func;
  private List<Type> paramTypes;
  private boolean hasExecutorHook;
  private Function<String, Object>[] paramParsers;
  private List<String> paramNames;
  private Function<String, Object> retValueParser;
  private BiPredicate<Object, Object> comparator;
  private boolean customExpectedType;

  /**
   * This constructor initializes type parsers for all arguments and return type
   * of func.
   *
   * @param func            - a method to test.
   * @param comparator   - an optional comparator for result. If comparator is
   *                     null, values are compared with equals().
   * @param expectedType - can be used with a custom comparator that has
   *                     different types for expected and result arguments.
   */
  public GenericTestHandler(Method func, BiPredicate<Object, Object> comparator,
                            List<Class<?>> expectedType) {
    this.func = func;
    this.comparator = comparator;

    hasExecutorHook = false;

    paramTypes = Arrays.asList(func.getGenericParameterTypes());
    if (paramTypes.size() >= 1 &&
        paramTypes.get(0).equals(TimedExecutor.class)) {
      hasExecutorHook = true;
      paramTypes = paramTypes.subList(1, paramTypes.size());
    }

    if (paramTypes.size() >= 1 && paramTypes.get(0).equals(TestTimer.class)) {
      throw new RuntimeException("This program uses deprecated TestTimer hook");
    }

    @SuppressWarnings("unchecked")
    Function<String, Object>[] A = new Function[paramTypes.size()];
    paramParsers = A;

    for (int i = 0; i < paramTypes.size(); i++) {
      paramParsers[i] = TestUtilsDeserialization.getTypeParser(
          TestUtilsDeserialization.linearizeType(paramTypes.get(i)));
    }

    paramNames = Arrays.stream(func.getParameters())
                     .map(Parameter::getName)
                     .collect(Collectors.toList());
    if (hasExecutorHook) {
      paramNames.remove(0);
    }

    if (expectedType == null) {
      retValueParser = TestUtilsDeserialization.getTypeParser(
          TestUtilsDeserialization.linearizeType(func.getGenericReturnType()));
    } else {
      retValueParser = TestUtilsDeserialization.getTypeParser(expectedType);
    }

    customExpectedType = expectedType != null;
  }

  /**
   * This method ensures that test data header matches with the signature
   * of the method provided in constructor.
   *
   * @param signature - the header from a test data file.
   */
  public void parseSignature(List<String> signature) {
    if (signature.size() != paramTypes.size() + 1) {
      throw new RuntimeException("Signature parameter count mismatch");
    }

    for (int i = 0; i < paramTypes.size(); i++) {
      if (!TestUtilsDeserialization.matchArgumentType(
              paramTypes.get(i),
              TestUtils.filterBracketComments(signature.get(i)))) {
        throw new RuntimeException(Integer.toString(i) +
                                   "th argument type mismatch");
      }
    }

    if (!customExpectedType) {
      if (!TestUtilsDeserialization.matchArgumentType(
              this.func.getGenericReturnType(),
              TestUtils.filterBracketComments(
                  signature.get(signature.size() - 1)))) {
        throw new RuntimeException("Return type mismatch");
      }
    }
  }

  /**
   * This method is invoked for each row in a test data file (except the
   * header). It deserializes the list of arguments and calls the user method
   * with them.
   *
   * @param testArgs - serialized arguments.
   * @return array, that contains [result of comparison of expected and result,
   * expected, result]. Two last entries are omitted in case of the void return
   * type
   */
  public TestTimer runTest(long timeoutSeconds, List<String> testArgs)
      throws Exception, Error {
    try {
      if (testArgs.size() !=
          paramParsers.length + (retValueParser != null ? 1 : 0)) {
        throw new RuntimeException("Invalid argument count");
      }

      List<Object> parsed = new ArrayList<>();
      for (int i = 0; i < paramParsers.length; i++) {
        parsed.add(paramParsers[i].apply(testArgs.get(i)));
      }

      Object result;
      TimedExecutor executor = new TimedExecutor(timeoutSeconds);

      if (hasExecutorHook) {
        parsed.add(0, executor);
        result = func.invoke(null, parsed.toArray());
      } else {
        result = executor.run(() -> func.invoke(null, parsed.toArray()));
      }

      if (!expectedIsVoid()) {
        Object expected =
            retValueParser.apply(testArgs.get(testArgs.size() - 1));
        assertResultsEqual(expected, result);
      }

      return executor.getTimer();
    } catch (IllegalAccessException e) {
      throw new RuntimeException(e.getMessage());
    } catch (InvocationTargetException e) {
      Throwable t = e.getCause();
      if (t instanceof Exception) {
        throw (Exception)t;
      } else if (t instanceof Error) {
        throw (Error) t;
      } else {
        // Improbable except for intended attempts to break the code, but anyway
        throw new RuntimeException(t);
      }
    }
  }

  private void assertResultsEqual(Object expected, Object result)
      throws TestFailure {
    boolean comparisonResult;
    if (comparator != null) {
      comparisonResult = comparator.test(expected, result);
    } else if (expected == null) {
      comparisonResult = result == null;
    } else if (expected instanceof Float && result instanceof Float) {
      comparisonResult =
          TestUtils.floatComparison((Float)expected, (Float)result);
    } else if (expected instanceof Double && result instanceof Double) {
      comparisonResult =
          TestUtils.doubleComparison((Double)expected, (Double)result);
    } else if (BinaryTreeUtils.isObjectTreeType(expected) ||
               BinaryTreeUtils.isObjectTreeType(result)) {
      BinaryTreeUtils.assertEqualBinaryTrees(expected, result);
      return;
    } else {
      comparisonResult = expected.equals(result);
    }
    if (!comparisonResult) {
      throw new TestFailure()
          .withProperty(TestFailure.PropertyName.EXPECTED, expected)
          .withProperty(TestFailure.PropertyName.RESULT, result);
    }
  }

  public boolean expectedIsVoid() { return retValueParser == null; }

  public List<String> paramNames() { return paramNames; }
}
